File: 202308/33.go
func search(nums []int, target int) int {
	lo, hi := 0, len(nums)-1
	for lo <= hi {
		mid := lo + (hi-lo)/2
		if target == nums[mid] {
			return mid
		} else if target == nums[lo] {
			return lo
		} else if target == nums[hi] {
			return hi
		} else {
			if nums[lo] < nums[mid] {
				if nums[lo] < target && target < nums[mid] {
					hi = mid - 1
				} else {
					lo = mid + 1
				}
			} else {
				if nums[mid] < target && target < nums[hi] {
					lo = mid + 1
				} else {
					hi = mid - 1
				}
			}
		}
	}
	return -1
}

File: 202605/33.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        lo, hi = 0, len(nums) - 1
        while lo <= hi:
            mi = lo + hi >> 1
            if nums[mi] == target:
                return mi
            if nums[lo] == target:
                return lo
            if nums[hi] == target:
                return hi
            elif nums[lo] < nums[mi]:
                if nums[lo] < target < nums[mi]:
                    hi = mi - 1
                else:
                    lo = mi + 1
            else:
                if nums[mi] < target < nums[hi]:
                    lo = mi + 1
                else:
                    hi = mi - 1
        return -1
